File: binary_expression_tree.py

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

    def print_in_order(self):
        if self.left:
            self.left.print_in_order()
        print(self.data)
        if self.right:
            self.right.print_in_order()


class BinaryExpressionTree:
    def __init__(self):
        self.stack = []
        self.operators = set(['+', '-', '*', '/', '(', ')', '^'])

    def construct_expression(self, expression):
        for i in range(len(expression)):
            node = Node(expression[i])
            if expression[i] not in self.operators:
                node.data = int(node.data)
                self.stack.append(node)
            else:
                r = self.stack.pop()
                l = self.stack.pop()
                node.right = r
                node.left = l
                self.stack.append(node)

    def evaluate(self, root):
        if root.data is None: # empty tree?
            return 0

        if root.left is None and root.right is None: # leaf?
            return root.data

        left_sum = self.evaluate(root.left) # evaluate left tree
        right_sum = self.evaluate(root.right) # evaluate right tree
        # well duh! there's no ^ evaluator here
        if root.data == '+':
            return left_sum + right_sum
        elif root.data == '-':
            return left_sum - right_sum
        elif root.data == '*':
            return left_sum * right_sum
        elif root.data == '^':
            return left_sum ** right_sum
        else:
            return left_sum // right_sum

            
test_tree = BinaryExpressionTree()
test_tree.construct_expression(['3', '5', '9', '+', '2', '*', '+'])
test_tree.stack[0].print_in_order()
answer = test_tree.evaluate(test_tree.stack[0])
print('answer: ', answer)
    
test_tree2 = BinaryExpressionTree()
test_tree2.construct_expression(['8', '2', '^', '6', '+', '2', '10', '*', '2', '/', '-'])
test_tree2.stack[0].print_in_order()
answer2 = test_tree2.evaluate(test_tree2.stack[0])
print('answer2: ', answer2)

File: BinaryExpressionTree.js
class Node {
    constructor(data) {
        this.data = data,
        this.left = null,
        this.right = null
    }
}


class BinaryExpressionTree {
    constructor() {
        this.stack = [],
        this.operations = {
            '^': (a, b) => a ** b,
            '*': (a, b) => a * b,
            '/': (a, b) => a / b,
            '+': (a, b) => a + b,
            '-': (a, b) => a - b
        } 
    }

    parse(expression) {
        expression.forEach((char) => {
            const node = new Node(char)
            if (!Object.keys(this.operations).includes(char)) {
                node.data = parseInt(node.data, 10)
                this.stack.push(node)
            } else {
                node.right = this.stack.pop()
                node.left = this.stack.pop()
                this.stack.push(node)
            }
        })    
    }

    operate(n1, o, n2)  { // operand in middle
        return this.operations[o](n1, n2) 
      }

    evaluate(root) {
        if (!root.data) { 
            return 0
        }
        if (!root.left && !root.right) {
            return root.data
        }
        const leftResult = this.evaluate(root.left)
        const rightResult = this.evaluate(root.right)
        return this.operate(leftResult, root.data, rightResult)
    }
    returnAnswer() {
        return this.evaluate(this.stack[0])
    }    
}


const input = ['8', '2', '^', '6', '+', '2', '10', '*', '2', '/', '-']
const testTree = new BinaryExpressionTree()
testTree.parse(input)
const answer = testTree.returnAnswer()
console.log('postfix input: ', input.join(' '))
console.log('answer: ', answer)
